xwiki-commons-core/xwiki-commons-xml/src/test/java/org/xwiki/xml/internal/html/DefaultHTMLCleanerTest.java

```diff
@@ -661,18 +664,104 @@ public void parseWithUnicodeChars() throws Exception
 
         assertEquals("baz&buz", nodeList.item(2).getAttributes().getNamedItem("foo").getTextContent());
 
-        // BUG: This is failing with baz&gt;buz
+        // This was failing with baz&gt;buz
         assertEquals("baz>buz", nodeList.item(1).getTextContent());
-        // BUG: This is failing with baz&amp;buz
+        // This was failing with baz&amp;buz
         assertEquals("baz&buz", nodeList.item(2).getTextContent());
     }
 
+    /**
+     * Test that line feed and carriage return are inserted in their unencoded form.
+     */
+    @Test
+    void doNotRemoveLFandCRInTextNodes()
+    {
+        assertHTML("<p>\r</p>", "<p>&#xD;</p>");
+        assertHTML("<p>Test \r\n</p>", "<p>Test \r\n</p>");
+    }
+
+    /**
+     * Ensure that &amp; in text nodes is not decoded even if it looks like it is following a character entity.
+     */
+    @Test
+    void doNotDecodeAmpersandInTextNodes()
+    {
+        assertHTML("<p>&amp;#xA;</p>", "<p>&amp;#xA;</p>");
+        assertHTML("<p>&amp;#10;</p>", "<p>&amp;#10;</p>");
+    }
+
+    /**
+     * Ensures that &lt; in text nodes is not decoded even if it is encoded as character entity.
+     */
+    @Test
+    void replaceOpeningBraceCharacterEntityByNamedEntity()
```

Comment:
Here I started writing even more tests to figure out how the 2.25 version behaves. These can be removed as well if the are only of personal interest for me than documenting wanted behavior (and possibly adding burden to future commits who need to keep these tests running).

Reply:
Never too many fast unit tests IMO. Make upgrades a lot safer.

Reply:
Agreed with Thomas.

Reply:
what's missing maybe is to explain why we want the behavior maybe...

Reply:
Actually in the case of the `replaceBrokenExplicitOpeningBraceByNamedEntity` I only document what the HtmlCleaner does, indeed. The input is invalid, so the HTML Cleaner has some leeway to decide what to create as output, because there is no unique solution.

If you say that tests only documenting the existing behavior are not good, I am ok with scrapping that test.

For the other test methods I will update the PR with a few more comments in the javadoc explaining why the test.